include/odp/api/spec/packet.h

```diff
@@ -83,6 +83,22 @@ typedef struct odp_packet_data_range {
 
 } odp_packet_data_range_t;
 
+/**
+ * Checksum check status in packet
+ */
+typedef enum odp_packet_chksum_status_t {
+	/** Checksum was not checked. Checksum check was not attempted or
+	  * the attempt failed. */
+	ODP_PACKET_CHKSUM_UNKNOWN = 0,
```

Comment:
OK and UNKNOWN are the most common cases: OK when checking is enabled, UNKNOWN when checking disabled. E.g. a router would not need check L4 checksum, so IPv4 check would be on, but all L4 checksum checks off.

Initialization is more robust with UNKNOWN == 0. Odp-linux would save these along the other packet flags (init would be flags.all_bits = 0). On the other hand, HW specific implementation may not store this enum anywhere, but directly mask HW specific packet descriptor flags to produce the status. In that case, the values do not matter too much.

Reply:
Then perhaps the 0 enum should be `ODP_PACKET_CHECKSUM_NORMAL` where "normal" is interpreted as OK if checksum offload was requested and "unknown" if it was not requested. The enum is then simply either NORMAL or BAD. The point is the only thing the application really wants to know is whether the checksum is bad or not. If it's disabled checksumming for whatever reason then presumably it doesn't care. 

If we reduce the enum to only two values, then it can simply collapse into a "bad" bit that's set if we know the checksum is incorrect and left as 0 otherwise. The 0 means it's OK if we are checking checksums and unknown if we don't care about them.

Reply:
These three values are needed. For example, we cannot dictate that all HW must be able to calculate checksums when there are extension headers. Still some HW can do that. So, even if application enabled checksumming, it depends on the packet and on the implementation if check was done or not. We just dictate that basic packets (no options / extensions) must always be checked when checksumming is enabled.

Reply:
About the values, we can change those later if there's evidence from various implementations that OK == 0 would be better than UNKNOWN == 0. If there's no measurable performance difference, UNKNOWN == 0 is more robust as init value.

Reply:
ODP APIs say nothing about what HW can or cannot do since the APIs don't presume any implementation model. They describe the services that ODP implementations provide in a platform-optimal manner. Ideally that means in HW, but if a given platform requires SW to provide that service in general, or in a specific case, then the implementation is in a better position to do that optimally than the application.

If the intent here is to request only services that are HW accelerated, then they should be recast in that light. But in the case of checksumming, it's not as if this can be skipped if it's "too hard" to do in HW. So I'd rather see ODP do it in SW once than requiring every application to duplicate this effort.

Reply:
The point is that we must be careful what is requested as a must from ODP (for every packet). E.g. ODP does not know on which packets application cares about checksum and on which not. If we require that ODP always checks checksum for all packets (also when HW cannot do it), an  implementation ends up routing packets to spare cores and back for SW checksum, which  application may very well ignore when received (since it was forwarding and not terminating those).

This API introduces the HW independent way to check if HW has checked the packet. We can add  full and partial checksum functions (to ODP API) for application assist, when HW check was not done for one reason or the other. This is optimal since not application calls for the SW checking only when it's really needed.